### PR TITLE
DE44059 - Implement fullscreen exit via esc

### PR DESCRIPTION
**Pull request by PaulKrznaric**
Users can now use the `esc` key to exit fullscreen view for the HTML Editor. 

Full Screen Esc Jiff

### CODE REVIEW

File: components/event-registry.js [hunk 1]
```diff
@@ -61,7 +61,7 @@ export class EventRegistry {
 			};
 		};
 
-		this.#editor.on('keydown', e => handler(trimEventDetails(e)));
+		this.#editor.on('keydown', e => handler(trimEventDetails(e), this.#editor.hasFocus()));
```
Comment (grant-cleary) [inline components/event-registry.js:64]: I _think_ I would prefer to see a separate `internalEditorHasFocus()` API in the `APIWrapper` class that exposes the latter instead. It seems a bit odd to return it from the event handler. Then the workflow would be to call `helpers.internalAPIWrapper.internalEditorHasFocus()` within the event handler to replace this second param. I want to think about this a bit more though, so maybe don't make any changes quite yet.

File: components/plugins/fullscreen.js [hunk 2]
```diff
@@ -15,30 +18,45 @@ export class FullscreenPlugin {
...
+			await helpers.internalAPIWrapper.toggleFullscreen(isFullscreen);
 			if (applyFocus) await helpers.internalAPIWrapper.setToolbarFocus('d2l-fullscreen');
 		};
 
-		const handleShortcutKey = e => {
-			if (e.keyCode !== 70) return;
-			if (!(e.metaKey && e.shiftKey) && !(e.ctrlKey && e.shiftKey)) return;
+		const handleShortcutKey = (e) => {
+			const isEditorFocused = helpers.internalAPIWrapper.internalEditorHasFocus();
+			const isValidEscEvent = (e.keyCode === escKey && isEditorFocused === true && isFullscreen === true);
```
Comment (grant-cleary) [inline components/plugins/fullscreen.js:47]:
```suggestion
			const isValidEscEvent = (e.keyCode === escKey && isEditorFocused && isFullscreen);
```

Comment (grant-cleary) [inline components/plugins/fullscreen.js:47]:
```suggestion
			const isValidEscEvent = (e.keyCode === escKey && isEditorFocused && isFullscreen);
```

File: components/plugins/fullscreen.js [hunk 3]
```diff
@@ -15,30 +18,45 @@ export class FullscreenPlugin {
...
+			const isValidFullscreenShortcut = e.keyCode === fKey && (e.metaKey && e.shiftKey) || (e.ctrlKey && e.shiftKey);
+			if (!isValidEscEvent && !isValidFullscreenShortcut) return;
+
 			toggleFullscreen(false);
+			if (e.keyCode !== escKey || isFullscreen === true) handleFullscreenDismiss();
```
Comment (grant-cleary) [inline components/plugins/fullscreen.js:52]:
```suggestion
			if (e.keyCode !== escKey || isFullscreen) handleFullscreenDismiss();
```

**Review by grant-cleary — APPROVED** [review]

File: components/plugins/fullscreen.js [hunk 4]
```diff
@@ -15,30 +18,45 @@ export class FullscreenPlugin {
...
 			if (applyFocus) await helpers.internalAPIWrapper.setToolbarFocus('d2l-fullscreen');
 		};
 
-		const handleShortcutKey = e => {
-			if (e.keyCode !== 70) return;
-			if (!(e.metaKey && e.shiftKey) && !(e.ctrlKey && e.shiftKey)) return;
+		const handleShortcutKey = (e) => {
+			const isEditorFocused = helpers.internalAPIWrapper.internalEditorHasFocus();
+			const isValidEscEvent = (e.keyCode === escKey && isEditorFocused && isFullscreen);
+			const isValidFullscreenShortcut = e.keyCode === fKey && (e.metaKey && e.shiftKey) || (e.ctrlKey && e.shiftKey);
```
Comment (klhuff) [inline components/plugins/fullscreen.js:48]: Should there be an extra pair of brackets around the pair of operators around the OR?

**Review by klhuff — APPROVED** [review]

File: components/plugins/fullscreen.js (hunk 4 above)
Comment (grant-cleary) [reply components/plugins/fullscreen.js:48]: Good catch, yes. `&&` has a higher operator precedence than `||`.

**Pull request merged**